### ISSUE TITLE
Bogus nullability warning for using lifted struct operators

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#72085]
**Version Used**: 
```
Branch main (19 Jan 2024)
Latest commit [d8d7886](https://github.com/dotnet/roslyn/commit/d8d7886808a68c720a246b09028f870bc62c30fe) by Jared Parsons:
Move projects to different directories (#71700)

* Move projects to different directories
The SDK / NuGet does not support projects existing in the same directory
during restore. It creates non-deterministic behavior. I noticed odd
behavior while tracking down a test failure in #71686. Digging into the
output it appeared our build was mixing outputs and steppping back I
noticed the double projects in the same directory.

This change updates BuildBoss to catch this mistake and refactors our
project layout to avoid the problem.

https://github.com/dotnet/sdk/issues/28763

* More
* More
```

**Steps to Reproduce**:

Compile the following code:
```cs
#nullable enable

LiftedOperatorUsage? s = new LiftedOperatorUsage();
s++;

public struct LiftedOperatorUsage
{
  public static LiftedOperatorUsage operator ++(LiftedOperatorUsage source) => source;
}
```

[sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4GIYFcA2OLA4CmABETAcQLABQtAMlAGYAuRCA8gA5EBOELAPa8AqgGcIAcyIB+EmJIBeEjCIB3Eo1btufAcPFSiACgCUAblpjkyS3RoABAMzyWvLAGMWm5m049+IVEJaVoAb1oSEmdogEYANh9tfz0gw2kSQQD9XhIbYy0/XUCDENIxQSxeDyJTJQA+eUrqojsAXyA==)

**Diagnostic Id**:

```
warning CS8620: Argument of type 'LiftedOperatorUsage?' cannot be used for parameter 'source' of type 'LiftedOperatorUsage' in 'LiftedOperatorUsage LiftedOperatorUsage.operator ++(LiftedOperatorUsage source)' due to differences in the nullability of reference types.
```

**Expected Behavior**:

No warnings.
The warning doesn't make sense as the operator is in fact never called with a nullable value, it's actually impossible due to its parameter having a struct type.
The compiler either skips the operator invocation completely if the value is nullable or unwarp the underlying struct to pass to the operator. Neither option calls the operator with a nullable value, so the warning is incorrect.

**Actual Behavior**:

Incorrect CS8620 warning is reported:
```
warning CS8620: Argument of type 'LiftedOperatorUsage?' cannot be used for parameter 'source' of type 'LiftedOperatorUsage' in 'LiftedOperatorUsage LiftedOperatorUsage.operator ++(LiftedOperatorUsage source)' due to differences in the nullability of reference types.
```

Might be related to https://github.com/dotnet/roslyn/issues/58201
